### PR TITLE
Add pre-commit hook and CI check, excluding currently unformatted files

### LINKED ISSUE #7634: pre-commit, editorconfig, nixpkgs-fmt?
**Is your feature request related to a problem? Please describe.**

Saw some unproductive comments about formatting today. Nothing too bad, but we could prevent this.

**Describe the solution you'd like**

- `nixpkgs-fmt`
- `pre-commit-hooks.nix` configuration that installs the hook in `shellHook`
- `.git-blame-ignore-revs` file, now also supported by GitHub web app

**Describe alternatives you've considered**

Carry on.

**Additional context**

I think we can just vote among the Nix Team and I'll be happy either way.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Eliminate discussions about formatting and make the code consistent, eventually. Make progress.

This will enforce formatting for ~30 files currently, and any new files added.

EDIT: 6 weeks later, I've had to add 14 new files to the ignore list.

# Context

**Implementation strategy**: minimally invasive addition; first turn off the faucet, then start drying.

**Inclusion of flake-parts**: the other entrypoints into pre-commit-hooks.nix are either non-flake (and a little odd), or don't support the system attributes we need.

 - Closes #7634 

 - https://github.com/NixOS/nix/pull/6721

# Scoped out

Currently the formatters only run manually. Pre-commit invocation depends on an upstream PR because we want it to be opt-in:

   - https://github.com/cachix/pre-commit-hooks.nix/pull/261

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [x] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

File: doc/manual/src/contributing/hacking.md
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/hacking.md:97]:
```suggestion
The hook configuration is only regenerated when you start a new development shell.
```

**Review by fricklerhandwerk — APPROVED** [review]
Not sure about including flake-parts, but I really like the incremental strategy which allows for deliberate changes that can be timed not to break conflicting PRs.

File: doc/manual/src/contributing/hacking.md
Comment (SuperSandro2000) [inline doc/manual/src/contributing/hacking.md:91]: This should be turned off by default. git hooks can be incredible annoying when they are not fully working and you are in the middle of a more complicated git action like a rebase or you are doing a commit which is broken on purpose and will be later fixed. Also when switching to an old branch when still inside the nix develop shell, this will probably blow up.

File: maintainers/flake-module.nix
Comment (SuperSandro2000) [inline maintainers/flake-module.nix:31]: I wouldn't mind if that would be reusable and also working for editorconfig but thats probably out of scope.

**Comment by edolstra** [comment]
I'm not in favor of doing this until #6721 is merged. I don't see a big advantage in applying formatting only to a couple of new files.

File: maintainers/flake-module.nix
Comment (roberth) [reply maintainers/flake-module.nix:34]: Editorconfig doesn't suffer from large diffs as much, I would think, so editorconfig won't really need this list. We could just apply its suggestions immediately.

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting 2023-03-13:

  - The pre-commit hook shouldn't be enabled by default (at least not for starters).
  - Create phony make target to manually run the formatter
  - Also try to unify the different `nixpkgs` dependencies in the Flake
  - Small issue around the use of `flake-parts` which shouldn't be blessed as the default way of doing things. But fairly minor, we can ignore it
  - Idea-approved as an experiment

File: doc/manual/src/contributing/hacking.md
Comment (9999years) [reply doc/manual/src/contributing/hacking.md:209]: I think it's OK. It's easy enough to use `git commit -n` to skip the Git hooks when needed.

It could be nice to have an environment variable to configure this behavior, like `NIX_DEVELOP_PRE_COMMIT`.

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-01-26.
Agreement on doing the CI+manual tooling part first.

<details>
- Blocked on https://github.com/cachix/pre-commit-hooks.nix/pull/261
- @thufschmitt: Can we skip the pre-commit-hook part at first?
    - @roberth: Would be very annoying
    - @ericson2314: Probably acceptable though
        - Would help building the case for the upstream PR
- What if we integrate it, but make it optional using different means
- @roberth: sgtm

**Comment by 9999years** [comment]
Why not just apply the `pre-commit` patch in our environment? Nix is very good at providing a patched dependency.

**Comment by roberth** [comment]
I've simplified the setup by making hook installation/update a manually invoked command.

While I would like for the config to be updated automatically when entering a dev shell, that's currently too complicated to pull off while also making it opt-in. It's not essential and could be added later.

Let's keep it simple for now and start automating away suggestions like https://github.com/NixOS/nix/pull/10537#pullrequestreview-2009430686.

**Comment by L-as** [comment]
Have you considered not using pre-commit? I don't see why such a large dependency is needed to put a script in .git/hooks/pre-commit.

**Comment by roberth** [comment]
> Have you considered not using pre-commit? I don't see why such a large dependency is needed to put a script in .git/hooks/pre-commit.

I have. Implementing a pre-commit hook correctly is actually non-trivial. It is a rather low level interface. I don't know all the intricacies, but iirc, one of the problems is that you will accidentally touch unstaged changes. Pre-commit takes care of that, and Nix and Nixpkgs handle dependencies quite well.

By duplicating this logic, we could indeed reduce our dependency footprint, but that would add undue maintenance cost that negates the benefit.
A better way to achieve a reduction in dependency footprint is to give cachix/git-hooks.nix a lightweight backend. That's compatible with pre-commit configuration items such as the `exclude` regex.

Yet another solution is to use a different solution just for `make format`.
I don't think this is a good idea, because a git hook is very useful, and we shouldn't have to maintain two configs.

File: doc/manual/src/contributing/hacking.md
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/hacking.md:276]:
```suggestion
## Formatting and pre-commit hooks
```

File: doc/manual/src/contributing/hacking.md
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/hacking.md:297]: Stop and start what?

File: doc/manual/src/contributing/hacking.md
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/hacking.md:295]: Which config? The dev shell?

File: doc/manual/src/contributing/hacking.md
Comment (fricklerhandwerk) [inline doc/manual/src/contributing/hacking.md:296]:
```suggestion
- if you use `make format`: exit the development shell and start it again by running `nix develop`.
```

File: flake.nix
Comment (fricklerhandwerk) [inline flake.nix:212]: What does that mean? The PR top post doesn't state the problem it's supposedly solving

**Review by fricklerhandwerk — APPROVED** [review]
Looks good in general, but needs more clarity for contributor documentation. I also suggest to add in-code comments that instruct contributors what to do and also say why that's important.

Generally, everyone: Please don't leave rationale only in the GitHub threads. Always put them into commit messages, contributor documentation, and in-code comments.

**Pull request merged**
